docs/tutorial/connection-monitor.md

```diff
@@ -0,0 +1,519 @@
+      return {
+        connectionMonitor: {
+          value: connectionMonitorService.state.value,
+          context: connectionMonitorService.state.context
+        }
+      }
+    }
+    case ACTION_TYPES.MAIN.TRIGGER_EVENT: {
+      connectionMonitorService.send(action.payload)
+      break
+    }
+    default: {
+      console.log(`Unrecognized action type: ${action.type}`)
+      break
+    }
+  }
+}
+
+ipcMain.handle(
+  CONNECTION_MONITOR_IPC_CHANNEL,
+  connectionMonitorMainProcessReducer
+)
+```
+
+### Renderer Process
+
+The final part of this example is the renderer process. As mentioned previously, it uses HTML, CSS, and ES2020 JavaScript and could be replaced by any frontend framework. The `mainAPI` defined in `preload.js` is available on the global `window` object and used to interact with the main process. The file has a collection of code blocks responsible for updating the UI in response to changes from the main process. The rendering is mainly controlled by two parts.
+
+First, the listener that utilizes the `mainAPI.addConnectionMonitorListener` method to sync with the main process. As defined in the [Context Bridge `mainAPI`](#context-bridge-mainapi) section, this method expects action objects and uses a `switch` statement to handle the different types. The `ERROR` type is used to send connection monitor error messages with the UI, and the `TRANSITION` type is used to initiate a page update when the state of either main process services change.
+
+```js
+function connectionMonitorListener (_, action) {
+  const { CONNECTION_MONITOR_IPC_REDUCER_ACTION_TYPES: ACTION_TYPES } =
+    window.mainAPI.constants
+  switch (action.type) {
+    case ACTION_TYPES.RENDERER.ERROR: {
+      updateError(action.payload)
+      break
+    }
+    case ACTION_TYPES.RENDERER.TRANSITION: {
+      updatePage(action.payload)
+      break
+    }
+    default:
+      console.log(`Unrecognized action ${action.type}`)
+  }
+}
+
+window.mainAPI.addConnectionMonitorListener(connectionMonitorListener)
+```
+
+Second, a `window.onload` method is defined to request the initial state of the main process services so the UI can render respectively. This function is important because the services may start executing before the page finishes loading. Additionally, this kind of method is helpful for fetching the initial state when only the renderer process is reloaded and the main process remains the same.
+
+```js
+window.onload = () => {
+  window.mainAPI.requestState().then((states) => {
+    updatePage(states)
+  })
+}
+```
+
+Finally, at the beginning of the renderer process, a couple of `onclick` handlers are set up for the respective buttons that are used to toggle the ping and authentication flows.
+
+```js
+document.getElementById('toggle-ping').onclick = function onPingButtonClick () {
+  window.mainAPI.togglePing()
+}
+
+document.getElementById('auth-button').onclick = function onAuthButtonClick () {
+  window.mainAPI.toggleAuth()
+}
+```
```

Comment:
Add a conclusion :)

Reply:
What are some things I should have in the conclusion?